src/security/ServerOptions.cc

```diff
@@ -390,12 +396,12 @@ Security::ServerOptions::loadDhParams()
     parsedDhParams.resetWithoutLocking(dhp);
 
 #else // OpenSSL 3.0+
-    const auto type = eecdhCurve.isEmpty() ? "DH" : "EC";
+    const auto type = "DH";
```

Comment:
> Parameters contained within the file pointed to by tls-dh= contains parameters used by either Elliptic-Curve DH cipher(s) or the classical DH cipher.

This is just not true, as I stated and shown in my introductory post. You could and we actually *have* configured squid to configure a ECDHE curve **and** DHE parameters with a group read from the file system. I've shown the relevant excerpt from the `openssl s_client` in my introductory post (`Server Tmp Key: ...`). This has been the behavior for years. And this was also the only behavior. There was no successful loading of ECDH parameters at all. 

Please refer to `src/security/ServerOptions.cc`. AFAICS, you have
- `parsedDhParams`, `dhParamsFile`
- `eecdhCurve` 

for `OPENSSL_MAJOR_VERSION < 3` you have
- `Security::ServerOptions::loadDhParams`: loads DHE Params from `dhParamsFile` and stores them into the `DH`-type struct `parsedDhParams`
- Security::ServerOptions::updateContextEecdh: derives nid + uses `SSL_CTX_set_tmp_ecdh` to set the ECDH curve to use; then calls `SSL_CTX_set_tmp_dh` if `parsedDhParams` is set
- unless I'm really mistaken, this code (as used for Squid versions 4 to 5) does never ever actually pass anything from the `dhParamsFile` that is related to ECDH

for the `#else` (openssl >= 3.x) branch:
- `Security::ServerOptions::loadDhParams`: is broken for `type = "EC"`, see above error
- `updateContextEecdh`: derives nid and uses `SSL_CTX_set1_groups` to set the curve for ECDH; this still works (otherwise, ECDHE would be completely broken) --- this is all that is ever needed to successfully setup ECDHE AFAICT
- `updateContextEecdh`: continues with still calling `SSL_CTX_set_tmp_dh(ctx.get(), parsedDhParams.get());`, which should always fail if you try to pass anything related to ECDHE

Thus, the PR tries to *restore* the bevahavior that Squid exhibited in the last major versions and that is aligned with reference documention:

> 	   tls-dh=[curve:]file
> 			File containing DH parameters for temporary/ephemeral DH key
> 			exchanges, optionally prefixed by a curve for ephemeral ECDH
> 			key exchanges.
> 			See OpenSSL documentation for details on how to create the
> 			DH parameter file. Supported curves for ECDH can be listed
> 			using the "openssl ecparam -list_curves" command.
> 			WARNING: EDH and EECDH ciphers will be silently disabled if
> 				 this option is not set.

Instead I'm now asked to cater for some envisioned behavior that does neither work in the current v6 branch (`SSL_CTX_set_tmp_dh` does **not** work with any `EVP_PKEY` structure, whether it contains additional ECDHE parameters or DHE parameters) nor is aligned with the actual reference documentation OR how squid behaved in the past. As far as I understand the openssl documentation with regard to that function, it has also never worked in the past, even if there would have been curves that took additional parameters.

I would argue that it does make much more sense to repair and restore documented and actual behavior exhibited by squid4 and 5 and if that loading of additional ECDHE parameters is an important feature treat it as such and implement it for a future version without breaking compatibility. 

Or phrased differently: How many users would gain anything right now from being able to load additional ECDHE parameters for those theoretically defined curves? 0, because nobody uses it at the moment, because it does not work anyway. How many would gain from restoring the documented and actual behavior? At least one :wave:, and potentially some more, who may not even know it because for TLS-1.3 or standard TLS-1.2, everything usually works.

Btw: citing [doc for openssl ecparam](https://docs.openssl.org/3.4/man1/openssl-ecparam/) 
> OpenSSL is currently not able to generate new groups and therefore this command can only create EC parameters from known (named) curves.

Reply:
> I'm now asked to cater for some envisioned behavior that does neither work in the current v6 branch nor is aligned with the actual reference documentation OR how squid behaved in the past.

I agree with many of your statements, but just want to clarify that we are currently in the process of reaching consensus of what should be changed (if anything). Since reviewers are not yet on the same page, I do not recommend making changes until there is consensus about the primary change request (at least).

The only change request were there seems to be consensus right now is replacing OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS with the old `0` value (even there, there is no consensus that zero is best; only potential consensus that polishing that parameter, if any polishing is needed, can be done in a separate PR!).